lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs

```diff
@@ -3888,46 +3864,74 @@ type RewardAccountBuilder k
         =  (k 'RootK XPrv, Passphrase "encryption")
         -> (XPrv, Passphrase "encryption")
 
-mkRewardAccountBuilder
-    :: forall ctx s k (n :: NetworkDiscriminant) shelley.
-        ( ctx ~ ApiLayer s k 'CredFromKeyK
-        , shelley ~ SeqState n ShelleyKey
-        , HardDerivation k
-        , Bounded (Index (AddressIndexDerivationType k) (AddressCredential k))
-        , WalletKey k
-        , Typeable s
-        , Typeable n
-        )
-    => ctx
+mkWithdrawal
+    :: forall s k (n :: NetworkDiscriminant)
+     . Typeable s
+    => Typeable n
+    => WorkerCtx (ApiLayer s k 'CredFromKeyK)
     -> WalletId
-    -> Maybe ApiWithdrawalPostData
-    -> Handler (Withdrawal, RewardAccountBuilder k)
-mkRewardAccountBuilder ctx wid withdrawal = do
-    let selfRewardCredentials (rootK, pwdP) =
-            (getRawKey $ deriveRewardAccount @k pwdP rootK, pwdP)
+    -> AnyCardanoEra
+    -> ApiWithdrawalPostData
+    -> Handler Withdrawal
+mkWithdrawal workerCtx walletId era apiWdrl =
+    case apiWdrl of
+        SelfWithdrawal ->
+            liftHandler $ mkSelfWithdrawal @s @k @n workerCtx walletId
+        ExternalWithdrawal mnemonic ->
+            liftHandler $ mkExternalWithdrawal workerCtx era mnemonic
+
+mkSelfWithdrawal
+    :: forall s k (n :: NetworkDiscriminant)
+     . Typeable s
+    => Typeable n
+    => WorkerCtx (ApiLayer s k 'CredFromKeyK)
+    -> WalletId
+    -> ExceptT ErrReadRewardAccount IO Withdrawal
+mkSelfWithdrawal workerCtx walletId = do
+    let netLayer = workerCtx ^. networkLayer
+        db = workerCtx ^. typed @(DBLayer IO s k)
+    (rewardAccount, _, derivationPath) <- W.readRewardAccount @s @k @n db walletId
+    withdrawalCoins <- liftIO $ getCachedRewardAccountBalance netLayer rewardAccount
```

Comment:
It looks like you dropped a `W.readNextWithdrawal` here, causing a withdrawal to be made regardless whether it's worth it or not.

Reply:
> withdrawalCoins

Avoids confusion with `wdrl :: Withdrawal` 👍 Though possibly something else like `accountBalance :: Coin` might be even clearer.